### PR TITLE
initial commit for potential energy law

### PR DESCRIPTION
**Pull request by cslade1970**

### CODE REVIEW

File: symplyphysics/laws/kinematic/potential_energy_from_mass_and_height.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    validate_input, validate_output, expr_to_quantity
+)
+
+# Description
+# Potential energy of body EP = m * g * h
+# where:
+# m - body mass
+# h - height from Earth surface
+# g - free fall acceleration
+potential_energy_of_body, free_fall_acceleration, height, body_mass = symbols('potential_energy_of_body free_fall_acceleration height body_mass')
```
Comment (blackyblack) [inline symplyphysics/laws/kinematic/potential_energy_from_mass_and_height.py:12]: Use `g` from sympy units module.

File: symplyphysics/laws/kinematic/potential_energy_from_mass_and_height.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+law = Eq(potential_energy_of_body, body_mass * free_fall_acceleration * height)
+
+def print():
+    return pretty(law, use_unicode=False)
+
+@validate_input(body_mass_=units.mass, free_fall_acceleration_=units.length/units.time**2, height_=units.length)
+@validate_output(units.energy)
+def calculate_potential_energy(body_mass_: Quantity, free_fall_acceleration_: Quantity, height_: Quantity) -> Quantity:
```
Comment (blackyblack) [inline symplyphysics/laws/kinematic/potential_energy_from_mass_and_height.py:20]: Do not pass free fall acceleration to this function.

File: examples/kinematic/potential_energy_from_mass_and_height.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+SI.set_quantity_scale_factor(m, 0.009 * units.kilogram)
+g = units.Quantity('g')
+SI.set_quantity_dimension(g, units.acceleration)
+SI.set_quantity_scale_factor(g, 9.82 * units.meter / units.second**2)
+h = units.Quantity('h')
+SI.set_quantity_dimension(h, units.length)
+SI.set_quantity_scale_factor(h, 1500 * units.meter)
+
+print("\nFormula is:\n{}".format(potential_energy.print()))
+result = potential_energy.calculate_potential_energy(m, g, h)
```
Comment (blackyblack) [inline examples/kinematic/potential_energy_from_mass_and_height.py:25]: Could we estimate somehow how dangerous is this amount of energy?

File: test/kinematic/potential_energy_from_mass_and_height_test.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    SI.set_quantity_dimension(h, units.length)
+    SI.set_quantity_scale_factor(h, 1 * units.meter)
+    Args = namedtuple('Args', ['m', 'g', 'h'])
+    return Args(m=m, g=g, h=h)
+
+def test_basic_energy(test_args):
+    result = potential_energy.calculate_potential_energy(test_args.m, test_args.g, test_args.h)
+    assert SI.get_dimension_system().equivalent_dims(result.dimension, units.energy)
+    result_energy = convert_to(result, units.joule).subs(units.joule, 1).evalf(3)
+    assert result_energy == approx(9.82, 0.01)
```
Comment (blackyblack) [inline test/kinematic/potential_energy_from_mass_and_height_test.py:27]: It seems you multiplied all values and put the result in the assertion. This kind of test does not actually test anything. I suggest to google for some real values that you can use in your test and verify it's correctness.

File: test/kinematic/potential_energy_from_mass_and_height_test.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+def test_bad_body_mass(test_args):
+    bm = units.Quantity('bm')
+    SI.set_quantity_dimension(bm, units.acceleration)
```
Comment (blackyblack) [inline test/kinematic/potential_energy_from_mass_and_height_test.py:31]: You set units to acceleration but use meters.

File: test/kinematic/potential_energy_from_mass_and_height_test.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    SI.set_quantity_scale_factor(bm, 1 * units.meter)
+    with raises(errors.UnitsError):
+        potential_energy.calculate_potential_energy(bm, test_args.g, test_args.h)
+    with raises(TypeError):
+        potential_energy.calculate_potential_energy(100, test_args.g, test_args.h)
+
+def test_bad_gravitation_acceleration(test_args):
+    bg = units.Quantity('bg')
+    SI.set_quantity_dimension(bg, units.length)
+    SI.set_quantity_scale_factor(bg, 1 * units.kilogram)
```
Comment (blackyblack) [inline test/kinematic/potential_energy_from_mass_and_height_test.py:41]: You set length but use kilogrammes.

File: test/kinematic/potential_energy_from_mass_and_height_test.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    with raises(errors.UnitsError):
+        potential_energy.calculate_potential_energy(test_args.m, bg, test_args.h)
+    with raises(TypeError):
+        potential_energy.calculate_potential_energy(test_args.m, 100, test_args.h)
+
+def test_bad_height(test_args):
+    bh = units.Quantity('bh')
+    SI.set_quantity_dimension(bh, units.acceleration)
+    SI.set_quantity_scale_factor(bh, 1 * units.meter / units.second**2)
```
Comment (blackyblack) [inline test/kinematic/potential_energy_from_mass_and_height_test.py:50]: You can use mass to shorten your code.

File: symplyphysics/laws/kinematic/potential_energy_from_mass_and_height.py (hunk 1 above)
Comment (cslade1970) [reply symplyphysics/laws/kinematic/potential_energy_from_mass_and_height.py:12]: может быть необходимо использовать acceleration_due_to_gravity ? g  указывает на gram меру массы

Comment (blackyblack) [reply symplyphysics/laws/kinematic/potential_energy_from_mass_and_height.py:12]: Yes, this is what I meant.

File: test/kinematic/potential_energy_from_mass_and_height_test.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
+from collections import namedtuple
+from pytest import approx, fixture, raises
+
+from symplyphysics import (
+    units, convert_to, SI, errors
+)
+from symplyphysics.laws.kinematic import potential_energy_from_mass_and_height as potential_energy
+
+# A body with a mass of 9 grams at an altitude of 1500 meters has a potential
+# energy of 132.57 joules.How much potential energy will remain in the body at
```
Comment (blackyblack) [inline test/kinematic/potential_energy_from_mass_and_height_test.py:10]:
```suggestion
# energy of 132.57 joules. How much potential energy will remain in the body at
```

File: test/kinematic/potential_energy_from_mass_and_height_test.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+# an altitude of 500 if the body began to fall at
+# an initial velocity of 0 m/s?
```
Comment (blackyblack) [inline test/kinematic/potential_energy_from_mass_and_height_test.py:12]: Remove new line: `# an altitude of 500 if the body began to fall at an initial velocity of 0 m/s?`

File: test/kinematic/potential_energy_from_mass_and_height_test.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    result = potential_energy.calculate_potential_energy(test_args.m, test_args.h)
+    assert SI.get_dimension_system().equivalent_dims(result.dimension, units.energy)
+    result_energy = convert_to(result, units.joule).subs(units.joule, 1).evalf(3)
+    assert result_energy == approx(44.19, 0.01)
+
+
+def test_bad_body_mass(test_args):
+    bm = units.Quantity('bm')
+    SI.set_quantity_dimension(bm, units.acceleration)
+    SI.set_quantity_scale_factor(bm, 1 * units.meter/units.second**2)
```
Comment (blackyblack) [inline test/kinematic/potential_energy_from_mass_and_height_test.py:35]: You can use length to shorten your code.

File: test/kinematic/potential_energy_from_mass_and_height_test.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+# A body with a mass of 9 grams at an altitude of 1500 meters has a potential
+# energy of 132.57 joules.How much potential energy will remain in the body at
+# an altitude of 500 if the body began to fall at
+# an initial velocity of 0 m/s?
+
+@fixture
+def test_args():
+    m = units.Quantity('m')
+    SI.set_quantity_dimension(m, units.mass)
+    SI.set_quantity_scale_factor(m, 0.009 * units.kilogram)
```
Comment (blackyblack) [inline test/kinematic/potential_energy_from_mass_and_height_test.py:18]: You can use grams here.

File: examples/kinematic/potential_energy_from_mass_and_height.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+from symplyphysics.laws.kinematic import potential_energy_from_mass_and_height as potential_energy
+
+#A bullet weighing 9 grams is fired from the gun vertically upwards and reaches
+#an altitude of 1,500 meters. Under the action of gravity it falls freely
+#to the ground and hits a shard of glass lying on the ground
+#Can a bullet break a piece of glass on the ground if it takes 100 joules of work ?
+
+m = units.Quantity('m')
+SI.set_quantity_dimension(m, units.mass)
+SI.set_quantity_scale_factor(m, 0.009 * units.kilogram)
```
Comment (blackyblack) [inline examples/kinematic/potential_energy_from_mass_and_height.py:13]: You can use grams here.

File: examples/kinematic/potential_energy_from_mass_and_height.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
+from symplyphysics import (
+    units, convert_to,solve, SI
+)
+from symplyphysics.laws.kinematic import potential_energy_from_mass_and_height as potential_energy
+
+#A bullet weighing 9 grams is fired from the gun vertically upwards and reaches
+#an altitude of 1,500 meters. Under the action of gravity it falls freely
+#to the ground and hits a shard of glass lying on the ground
+#Can a bullet break a piece of glass on the ground if it takes 100 joules of work ?
```
Comment (blackyblack) [inline examples/kinematic/potential_energy_from_mass_and_height.py:9]: Breaking glass does not depend on work. It depends of pressure, I guess.
I'll think how to improve this example.

Comment (cslade1970) [reply examples/kinematic/potential_energy_from_mass_and_height.py:9]: For creating pressure need some work, really?

Comment (blackyblack) [reply examples/kinematic/potential_energy_from_mass_and_height.py:9]: Yes, but I guess it depends on force. Work is more about dissipating heat or moving things. Breaking is another process.

File: examples/kinematic/potential_energy_from_mass_and_height.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+# Under what condition will the potential energies of two different bodies be the same
+# if the mass of the first body is half the mass of the second body?
+# h1 - height body 1
+# h2 - height body 2
+# m1 - mass body 1
+# m2 - mass body 2
+# m2 = 2 * m1
+# Ep1 = Ep2
+m1, m2, h1, h2 = symbols('m1 m2 h1 h2')
```
Comment (blackyblack) [inline examples/kinematic/potential_energy_from_mass_and_height.py:17]: You can use full names for mass and height and remove comments about them.

Comment (blackyblack) [inline examples/kinematic/potential_energy_from_mass_and_height.py:17]: It seems m2 is not used.

File: examples/kinematic/potential_energy_from_mass_and_height.py [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+Ep1 = potential_energy.law.subs(potential_energy.body_mass, m1).subs(potential_energy.height,h1)
```
Comment (blackyblack) [inline examples/kinematic/potential_energy_from_mass_and_height.py:18]: Use subs({mass: m1, height: h1}) for shorter code.

File: examples/kinematic/potential_energy_from_mass_and_height.py [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+from symplyphysics.laws.kinematic import potential_energy_from_mass_and_height as potential_energy
+
+# Under what condition will the potential energies of two different bodies be the same
+# if the mass of the first body is half the mass of the second body?
+# h1 - height body 1
+# h2 - height body 2
+# m1 - mass body 1
+# m2 - mass body 2
+# m2 = 2 * m1
+# Ep1 = Ep2
```
Comment (blackyblack) [inline examples/kinematic/potential_energy_from_mass_and_height.py:16]: Remove this comment. It should be obvious from the code.

File: examples/kinematic/potential_energy_from_mass_and_height.py [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+m1, m2, h1, h2 = symbols('m1 m2 h1 h2')
+Ep1 = potential_energy.law.subs(potential_energy.body_mass, m1).subs(potential_energy.height,h1)
+Ep2 = potential_energy.law.subs(potential_energy.body_mass, 2 * m1).subs(potential_energy.height,h2)
+law = Eq(Ep2, Ep1)
+solved = solve(law, h1, dict=True)[0][h1]
+answer = Eq(h1,solved)
+print("\nFormula is:\n{}".format(potential_energy.print()))
```
Comment (blackyblack) [inline examples/kinematic/potential_energy_from_mass_and_height.py:23]: Do not forget spaces after comma. Here and everywhere.

File: test/kinematic/potential_energy_from_mass_and_height_test.py [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    SI.set_quantity_scale_factor(h, 500 * units.meter)
+    Args = namedtuple('Args', ['m', 'h'])
+    return Args(m=m, h=h)
+
+def test_basic_energy(test_args):
+    result = potential_energy.calculate_potential_energy(test_args.m, test_args.h)
+    assert SI.get_dimension_system().equivalent_dims(result.dimension, units.energy)
+    result_energy = convert_to(result, units.joule).subs(units.joule, 1).evalf(3)
+    assert result_energy == approx(44.19, 0.01)
+
```
Comment (blackyblack) [inline test/kinematic/potential_energy_from_mass_and_height_test.py:28]: Remove empty line.

**Review by blackyblack — APPROVED** [review]

File: examples/kinematic/potential_energy_from_mass_and_height.py [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+# Under what condition will the potential energies of two different bodies be the same
+# if the mass of the first body is half the mass of the second body?
+body_mass_1, height_1, height_2 = symbols('body_mass_1 height_1 height_2')
+Ep1 = potential_energy.law.subs({potential_energy.body_mass: body_mass_1, potential_energy.height: height_1})
+Ep2 = potential_energy.law.subs({potential_energy.body_mass: 2 * body_mass_1, potential_energy.height: height_2})
+law = Eq(Ep2, Ep1)
+solved = solve(law, height_1, dict=True)[0][height_1]
+answer = Eq(height_1, solved)
+print("\nFormula is:\n{}".format(potential_energy.print()))
+print("\n IF potential_energy_1 = potential_energy_2 AND body_mass_2 = 2 * body_mass_1 THEN ",
```
Comment (blackyblack) [inline examples/kinematic/potential_energy_from_mass_and_height.py:18]:
```suggestion
print("\nSolution:\nIF potential_energy_1 = potential_energy_2 AND body_mass_2 = 2 * body_mass_1 THEN {}".format(pretty(answer, use_unicode=False)))
```

**Review by blackyblack — APPROVED** [review]

**Pull request merged**